File: packages/go-agentenkit/usage_currency_test.go
package agentenkit_test

import (
	"testing"

	"github.com/zendev-sh/goai/provider"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// Workstream H: money is never converted. CostMicros is one currency (the
// first seen); a call priced in another is left out of it and counted as
// unpriced, and Costs keeps each currency's own total. The same cases run in
// the TS package (test/usage-currency.test.ts).

func TestUsageAggregatorKeepsOneTotalPerCurrency(t *testing.T) {
	var agg ports.UsageAggregator
	agg.Add(ports.NewUsage{AgentName: "a", Model: "m", InputTokens: 10, Cost: &ports.Cost{Micros: 100, Currency: "USD"}})
	agg.Add(ports.NewUsage{AgentName: "a", Model: "m", InputTokens: 10, Cost: &ports.Cost{Micros: 900, Currency: "EUR"}})
	agg.Add(ports.NewUsage{AgentName: "a", Model: "m", InputTokens: 10})
	total := agg.Totals()
	mustEqual(t, total.Currency, "USD", "the first currency seen")
	mustEqual(t, total.CostMicros, int64(100), "only that currency is summed")
	mustEqual(t, total.Unpriced, 2, "one unpriced, one in another currency")
	mustEqual(t, len(total.Costs), 2, "one total per currency")
	mustEqual(t, total.Costs[1], ports.CurrencyCost{Currency: "EUR", CostMicros: 900, Calls: 1}, "the euros are kept")
	mustEqual(t, len(total.Lines), 2, "a line per agent, model and currency")
	mustEqual(t, total.Lines[0].Calls, 2, "the unpriced call joins the first line")
	mustEqual(t, total.Lines[1].Currency, "EUR", "the euro line says so")
}

// The SQL adapters read GROUP BY agent, model and currency; the merger folds
// those groups back the same way.
func TestUsageLineMergerKeepsOneTotalPerCurrency(t *testing.T) {
	var m ports.UsageLineMerger
	line := func(cost int64, calls int) ports.UsageLine {
		return ports.UsageLine{AgentName: "a", Model: "m", InputTokens: 5, Calls: calls, CostMicros: cost}
	}
	m.Add(line(0, 1), "", 5, 1) // unpriced rows first: the line takes a currency later
	m.Add(line(100, 2), "USD", 5, 0)
	m.Add(line(700, 1), "EUR", 5, 0)
	total := m.Totals()
	mustEqual(t, total.Currency, "USD", "the first currency seen")
	mustEqual(t, total.CostMicros, int64(100), "only that currency is summed")
	mustEqual(t, total.Unpriced, 2, "one unpriced, one in another currency")
	mustEqual(t, total.TotalTokens, 15, "every call's tokens")
	mustEqual(t, len(total.Lines), 2, "a line per agent, model and currency")
	mustEqual(t, total.Lines[0].Currency, "USD", "the unpriced line took the first currency")
	mustEqual(t, total.Lines[0].Calls, 3, "three calls on it")
	mustEqual(t, total.Costs[1].CostMicros, int64(700), "the euros are kept")
}

// Total tokens is input + cached + output for every provider, never the
// provider's own total (Anthropic's leaves the cache reads out).
func TestUsage_TotalTokensIsAlwaysTheSum(t *testing.T) {
	got := agentenkit.AttributeTokens(provider.Usage{InputTokens: 10, CacheReadTokens: 30, OutputTokens: 5, TotalTokens: 15})
	mustEqual(t, got.TotalTokens, 45, "input + cached + output")
}

File: packages/agentenkit/test/usage-currency.test.ts
import { describe, expect, it } from 'bun:test';
import { attributeTokens, sumUsage, UsageMerger } from '../src/core/usage.js';
import type { NewUsage, UsageLine } from '../src/core/types.js';

// Workstream H: money is never converted. costMicros is one currency (the
// first seen); a call priced in another is left out of it and counted as
// unpriced, and costs keeps each currency's own total. The same cases run in
// the Go package (usage_currency_test.go).

const row = (cost?: { micros: number; currency: string }): NewUsage => ({
  agentName: 'a', model: 'm', kind: 'step', step: 1, outcome: 'finished',
  inputTokens: 10, cacheReadInputTokens: 0, cacheWriteInputTokens: 0, outputTokens: 0,
  reasoningTokens: 0, totalTokens: 10,
  ...(cost ? { cost: { ...cost, source: 'table' as const } } : {}),
});

describe('usage totals by currency (§4)', () => {
  it('the aggregator keeps one total per currency', () => {
    const total = sumUsage([row({ micros: 100, currency: 'USD' }), row({ micros: 900, currency: 'EUR' }), row()]);
    expect(total.currency).toBe('USD'); // the first currency seen
    expect(total.costMicros).toBe(100); // only that currency is summed
    expect(total.unpriced).toBe(2); // one unpriced, one in another currency
    expect(total.costs).toHaveLength(2);
    expect(total.costs![1]).toEqual({ currency: 'EUR', costMicros: 900, calls: 1 });
    expect(total.lines).toHaveLength(2); // a line per agent, model and currency
    expect(total.lines[0]!.calls).toBe(2); // the unpriced call joins the first line
    expect(total.lines[1]!.currency).toBe('EUR');
  });

  it('the line merger keeps one total per currency', () => {
    const m = new UsageMerger();
    const line = (costMicros: number, calls: number): UsageLine => ({
      agentName: 'a', model: 'm', inputTokens: 5, cacheReadInputTokens: 0, cacheWriteInputTokens: 0,
      outputTokens: 0, reasoningTokens: 0, calls, estimated: 0, costMicros,
    });
    m.add({ line: line(0, 1), currency: null, totalTokens: 5, unpriced: 1 }); // the line takes a currency later
    m.add({ line: line(100, 2), currency: 'USD', totalTokens: 5, unpriced: 0 });
    m.add({ line: line(700, 1), currency: 'EUR', totalTokens: 5, unpriced: 0 });
    const total = m.totals();
    expect(total.currency).toBe('USD');
    expect(total.costMicros).toBe(100);
    expect(total.unpriced).toBe(2);
    expect(total.totalTokens).toBe(15);
    expect(total.lines).toHaveLength(2);
    expect(total.lines[0]!.currency).toBe('USD'); // the unpriced line took the first currency
    expect(total.lines[0]!.calls).toBe(3);
    expect(total.costs![1]!.costMicros).toBe(700);
  });
});

describe('total tokens (§4)', () => {
  it('total tokens is always the sum', () => {
    // Anthropic's own total leaves the cache reads out; the count must not.
    const got = attributeTokens(
      { inputTokens: 10, outputTokens: 5, totalTokens: 15 },
      { anthropic: { cacheReadInputTokens: 30 } },
    );
    expect(got.totalTokens).toBe(45);
  });
});
